lib/Sema/TypeCheckConstraints.cpp

```diff
@@ -3323,6 +3323,10 @@ bool TypeChecker::convertToType(Expr *&expr, Type type, DeclContext *dc,
   // TODO: need to add kind arg?
   // Construct a constraint system from this expression.
   ConstraintSystem cs(*this, dc, ConstraintSystemFlags::AllowFixes);
+  
+  // We need to walk through the expression to set the type variables.
+  cs.generateConstraints(expr);
```

Comment:
This also hits at the symptoms but not the cause.  You're on the right track that `convertToType` is being used incorrectly.  It currently has three callers, and of those only the one in `TypeCheckStmt` has a valid reason to exist.  The remaining two can be replaced by calls to `Solution::coerceToType`.